Objects/genobject.c

```diff
@@ -282,46 +273,101 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
     }
 
     assert(gen->gi_exc_state.exc_value == NULL);
-    assert(gen->gi_frame_state == FRAME_CLEARED);
+    assert(frame_state == FRAME_CLEARED);
     *presult = result;
     return result ? PYGEN_RETURN : PYGEN_ERROR;
 }
 
+// Set the generator 'gen' to the executing state and send 'arg' into it.
+// See gen_send_ex2() for details.
+static PySendResult
+gen_send_ex(PyGenObject *gen, PyObject *arg, PyObject **presult)
+{
+    *presult = NULL;
+    int8_t frame_state = FT_ATOMIC_LOAD_INT8_RELAXED(gen->gi_frame_state);
+retry:
+    if (frame_state == FRAME_CREATED && arg && arg != Py_None) {
+        const char *msg = "can't send non-None value to a "
+                            "just-started generator";
+        if (PyCoro_CheckExact(gen)) {
+            msg = NON_INIT_CORO_MSG;
+        }
+        else if (PyAsyncGen_CheckExact(gen)) {
+            msg = "can't send non-None value to a "
+                    "just-started async generator";
+        }
+        PyErr_SetString(PyExc_TypeError, msg);
+        return PYGEN_ERROR;
+    }
+    if (frame_state == FRAME_EXECUTING) {
+        gen_raise_already_executing_error(gen);
+        return PYGEN_ERROR;
+    }
+    if (FRAME_STATE_FINISHED(frame_state)) {
+        if (PyCoro_CheckExact(gen)) {
+            /* `gen` is an exhausted coroutine: raise an error,
+               except when called from gen_close(), which should
+               always be a silent method. */
+            PyErr_SetString(
+                PyExc_RuntimeError,
+                "cannot reuse already awaited coroutine");
+        }
+        else if (arg) {
+            /* `gen` is an exhausted generator:
+               only return value if called from send(). */
+            *presult = Py_NewRef(Py_None);
+            return PYGEN_RETURN;
+        }
+        return PYGEN_ERROR;
+    }
+
+    assert((frame_state == FRAME_CREATED) ||
+           FRAME_STATE_SUSPENDED(frame_state));
+
+    if (!_Py_GEN_SET_FRAME_STATE(gen, frame_state, FRAME_EXECUTING)) {
+        goto retry;
```

Comment:
No need for a goto. You can use `do ... while (!_Py_GEN_SET_FRAME_STATE(gen, frame_state, FRAME_EXECUTING));`